### PR TITLE
Python: More Nullptr Checks

### PR DESCRIPTION
**Pull request by ax3l**
I am debugging a couple of crashes while mitigating CI.
One of them pointed to the locations here (`ValueError: NULL pointer access`), so I decided to add some checks.

### CODE REVIEW

**Review by roelof-groenewald — APPROVED** [review]
Looks good to me! There are also the functions `get_particle_structs`  and `get_particle_boundary_buffer_structs` that could use the same checks.

**Comment by ax3l** [comment]
Thanks, good idea. Let's add this there as well.

I think with this check, `Python_particle_attr_access` is actually crashing right now. I was wondering if it's a dormant bug or just my local setup, but now we got the finger on it :) Possibly depends on the version of `ctypes`/`numpy` that is used if this is silent UB, ignored or a problem.

**Comment by roelof-groenewald** [comment]
Just as a reference: the test passes locally for me, running with `numpy==1.21.0` and `ctypes==1.1.0`. It looks like the Azure test was run with `numpy==1.21.4`.

File: Python/pywarpx/_libwarpx.py [hunk 1]
```diff
@@ -599,6 +599,8 @@ def get_particle_arrays(species_name, comp_name, level):
 
     particle_data = []
     for i in range(num_tiles.value):
+        if not data[i]:
+            raise Exception(f'get_particle_arrays: data[i] for i={i} was not initialized')
```
Comment (ax3l) [inline Python/pywarpx/_libwarpx.py:609]:
```suggestion
            continue
```

File: Python/pywarpx/_libwarpx.py [hunk 2]
```diff
@@ -949,6 +951,8 @@ def get_particle_boundary_buffer(species_name, boundary, comp_name, level):
...
```
Comment (ax3l) [inline Python/pywarpx/_libwarpx.py:965]:
```suggestion
            continue
```

File: Python/pywarpx/_libwarpx.py [hunk 3]
```diff
@@ -989,6 +993,8 @@ def _get_mesh_field_list(warpx_func, level, direction, include_ghosts):
     for i in range(size.value):
         shape = tuple([shapes[shapesize*i + d] for d in range(shapesize)])
         # --- The data is stored in Fortran order, hence shape is reversed and a transpose is taken.
+        if not data[i]:
+            raise Exception(f'get_particle_arrays: data[i] for i={i} was not initialized')
```
Comment (ax3l) [inline Python/pywarpx/_libwarpx.py:1009]:
```suggestion
            continue
```

**Review by ax3l — COMMENTED** [review]
Let's try to simply skip empty tiles

File: Python/pywarpx/_libwarpx.py [hunk 4]
```diff
@@ -599,6 +599,8 @@ def get_particle_arrays(species_name, comp_name, level):
 
     particle_data = []
     for i in range(num_tiles.value):
+        if not data[i]:
+            continue
```
Comment (ax3l) [inline Python/pywarpx/_libwarpx.py]:
```suggestion
        if particles_per_tile[i] == 0:
            continue
        if not data[i]:
            raise Exception(f'get_particle_arrays: data[i] for i={i} was not initialized')
```

File: Python/pywarpx/_libwarpx.py [hunk 5]
```diff
@@ -949,6 +951,8 @@ def get_particle_boundary_buffer(species_name, boundary, comp_name, level):
...
```
Comment (ax3l) [inline Python/pywarpx/_libwarpx.py]:
```suggestion
        if particles_per_tile[i] == 0:
            continue
        if not data[i]:
            raise Exception(f'get_particle_arrays: data[i] for i={i} was not initialized')
```

File: Python/pywarpx/_libwarpx.py [hunk 6]
```diff
@@ -989,6 +993,8 @@ def _get_mesh_field_list(warpx_func, level, direction, include_ghosts):
     for i in range(size.value):
         shape = tuple([shapes[shapesize*i + d] for d in range(shapesize)])
         # --- The data is stored in Fortran order, hence shape is reversed and a transpose is taken.
+        if not data[i]:
+            continue
```
Comment (ax3l) [inline Python/pywarpx/_libwarpx.py]:
```suggestion
        if shape[::-1] == 0:
            continue
        if not data[i]:
            raise Exception(f'get_particle_arrays: data[i] for i={i} was not initialized')
```

**Comment by roelof-groenewald** [comment]
Looks like that fixed it!

**Comment by ax3l** [comment]
Yay! Will add in those extra functions you mentioned to finalize:
https://github.com/ECP-WarpX/WarpX/pull/2574#pullrequestreview-809187849

**Comment by ax3l** [comment]
Fancy (unrelated) CI fails in `cartesian_3D`:
```
working on test: reduced_diags_loadbalancecosts_timers_psatd
   re-making clean...
   building...
   found pre-built executable for this test
   copying files to run directory...
   path to input file: Examples/Tests/reduced_diags/inputs_loadbalancecosts
   running the test...
   mpiexec -n 2 ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex inputs_loadbalancecosts  diag1.file_prefix=reduced_diags_loadbalancecosts_timers_psatd_plt   amrex.abort_on_unused_inputs=1 amrex.fpe_trap_invalid=1 amrex.fpe_trap_zero=1 amrex.fpe_trap_overflow=1 warpx.do_dynamic_scheduling=0 warpx.serialize_ics=1 algo.load_balance_costs_update=Timers
   WARNING: unable to open the job_info file
   doing the analysis...
   WARNING: analysis failed...
./analysis_reduced_diags_loadbalancecosts.py  reduced_diags_loadbalancecosts_timers_psatd_plt00003
load balance efficiency (before load balance):  0.9548868098837517
load balance efficiency (after load balance):  0.9196151100195358
Traceback (most recent call last):
  File "./analysis_reduced_diags_loadbalancecosts.py", line 69, in <module>
    assert(efficiency_before < efficiency_after)
AssertionError

   execution time: 1.6s
   creating problem test report ...
   reduced_diags_loadbalancecosts_timers_psatd FAILED
```
and
```
 5: /lib/x86_64-linux-gnu/libstdc++.so.6(+0x9e911) [0x7fd579967911]
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_data(char*) at /usr/include/c++/9/bits/basic_string.h:179
 (inlined by) void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) at /usr/include/c++/9/bits/basic_string.tcc:219
 (inlined by) void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct_aux<char const*>(char const*, char const*, std::__false_type) at /usr/include/c++/9/bits/basic_string.h:247

 6: /lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa38c) [0x7fd57997338c]
    amrex::Box::coarsen(amrex::IntVect const&) at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_Box.H:801
 (inlined by) PEC::ApplyPECtoBfield(std::array<amrex::MultiFab*, 3ul>, int, PatchType) at /tmp/ci-oFCLwlOx0A/warpx/./Source/BoundaryConditions/WarpX_PEC.cpp:131

 7: /lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa3f7) [0x7fd5799733f7]
    amrex::coarsen(int, int) at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_IntVect.H:28
 (inlined by) amrex::IntVect::coarsen(amrex::IntVect const&) at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_IntVect.H:591
 (inlined by) amrex::Box::coarsen(amrex::IntVect const&) at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_Box.H:804
 (inlined by) PEC::ApplyPECtoBfield(std::array<amrex::MultiFab*, 3ul>, int, PatchType) at /tmp/ci-oFCLwlOx0A/warpx/./Source/BoundaryConditions/WarpX_PEC.cpp:131

 8: /lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa6a9) [0x7fd5799736a9]
    amrex::coarsen(int, int) at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_IntVect.H:32
 (inlined by) amrex::IntVect::coarsen(amrex::IntVect const&) at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_IntVect.H:591
 (inlined by) amrex::Box::coarsen(amrex::IntVect const&) at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_Box.H:804
 (inlined by) PEC::ApplyPECtoBfield(std::array<amrex::MultiFab*, 3ul>, int, PatchType) at /tmp/ci-oFCLwlOx0A/warpx/./Source/BoundaryConditions/WarpX_PEC.cpp:131

 9: /lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt19__throw_ios_failurePKc+0x91) [0x7fd57996ac23]
    ?? ??:0

10: /lib/x86_64-linux-gnu/libstdc++.so.6(+0x114b72) [0x7fd5799ddb72]
    amrex::PODVector<int, std::allocator<int> >::GetNewCapacity(unsigned long) const at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_PODVector.H:504
 (inlined by) amrex::PODVector<int, std::allocator<int> >::resize(unsigned long) at /tmp/ci-oFCLwlOx0A/amrex//Src/Base/AMReX_PODVector.H:445
 (inlined by) amrex::ParticleContainer<0, 0, 4, 0, amrex::PinnedArenaAllocator>::SetParticleSize() at /tmp/ci-oFCLwlOx0A/amrex//Src/Particle/AMReX_ParticleContainerI.H:9

11: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0xfb4a8) [0x55f88d9534a8]
    std::basic_ios<char, std::char_traits<char> >::setstate(std::_Ios_Iostate) at /usr/include/c++/9/bits/basic_ios.h:158
 (inlined by) std::basic_ifstream<char, std::char_traits<char> >::open(char const*, std::_Ios_Openmode) at /usr/include/c++/9/fstream:661
 (inlined by) std::basic_ifstream<char, std::char_traits<char> >::open(char const*, std::_Ios_Openmode) at /usr/include/c++/9/fstream:658
 (inlined by) BTDPlotfileHeaderImpl::ReadHeaderData() at /tmp/ci-oFCLwlOx0A/warpx/./Source/Diagnostics/BTD_Plotfile_Header_Impl.cpp:32

12: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0xef930) [0x55f88d947930]
    BTDPlotfileHeaderImpl::set_timestep(int) at /tmp/ci-oFCLwlOx0A/warpx/./Source/Diagnostics/BTD_Plotfile_Header_Impl.H:83
 (inlined by) BTDiagnostics::InterleaveBufferAndSnapshotHeader(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) at /tmp/ci-oFCLwlOx0A/warpx/./Source/Diagnostics/BTDiagnostics.cpp:739

13: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0xf21b2) [0x55f88d94a1b2]
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_is_local() const at /usr/include/c++/9/bits/basic_string.h:222
 (inlined by) std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose() at /usr/include/c++/9/bits/basic_string.h:231
 (inlined by) std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() at /usr/include/c++/9/bits/basic_string.h:658
 (inlined by) BTDiagnostics::MergeBuffersForPlotfile(int) at /tmp/ci-oFCLwlOx0A/warpx/./Source/Diagnostics/BTDiagnostics.cpp:712

14: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0xf2c26) [0x55f88d94ac26]
    BTDiagnostics::Flush(int) at /tmp/ci-oFCLwlOx0A/warpx/./Source/Diagnostics/BTDiagnostics.cpp:649

15: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0xae888) [0x55f88d906888]
    Diagnostics::FilterComputePackFlush(int, bool) at /tmp/ci-oFCLwlOx0A/warpx/./Source/Diagnostics/Diagnostics.cpp:341

16: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0xad201) [0x55f88d905201]
    MultiDiagnostics::FilterComputePackFlush(int, bool) at /tmp/ci-oFCLwlOx0A/warpx/./Source/Diagnostics/MultiDiagnostics.cpp:74 (discriminator 2)

17: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0x437584) [0x55f88dc8f584]
    WarpX::Evolve(int) at /tmp/ci-oFCLwlOx0A/warpx/./Source/Evolve/WarpXEvolve.cpp:341

18: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0x5e854) [0x55f88d8b6854]
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Alloc_hider::_Alloc_hider(char*, std::allocator<char> const&) at /usr/include/c++/9/bits/basic_string.h:157
 (inlined by) std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&) at /usr/include/c++/9/bits/basic_string.h:526
 (inlined by) main at /tmp/ci-oFCLwlOx0A/warpx/./Source/main.cpp:69

19: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7fd57952d0b3]

20: ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex(+0x6eb3e) [0x55f88d8c6b3e]
    ?? ??:0


===== TinyProfilers ======
main()
WarpX::Evolve()
WarpX::Evolve::step
Diagnostics::FilterComputePackFlush()


   WARNING: +++ End of backtrace: BTD_ReducedSliceDiag.Backtrace.0.0 +++

   BTD_ReducedSliceDiag CRASHED (backtraces produced)
```

**Comment by ax3l** [comment]
Thx for the review @roelof-groenewald :)

**Pull request merged**